rts/Rendering/GL/State.h

```diff
@@ -0,0 +1,182 @@
+
+	ATTRIBUTE_TYPE_DEFS            (LineWidth, GL_LINE_WIDTH);
+
+	ATTRIBUTE_TYPE_DEFS            (Viewport, GL_VIEWPORT);
+
+	ATTRIBUTE_TYPE_DEFS            (FrontFace, GL_FRONT_FACE);
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (Culling, GL_CULL_FACE);
+	ATTRIBUTE_TYPE_DEFS            (CullFace, GL_CULL_FACE_MODE);
+	MULTI_CAPABILITY_ATTRIBUTE_TYPE_DEFS (ClipDistance, GL_CLIP_DISTANCE0);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (ScissorTest, GL_SCISSOR_TEST);
+	ATTRIBUTE_TYPE_DEFS            (Scissor, GL_SCISSOR_BOX);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (DepthTest, GL_DEPTH_TEST);
+	ATTRIBUTE_TYPE_DEFS            (DepthFunc, GL_DEPTH_FUNC);
+	ATTRIBUTE_TYPE_DEFS            (DepthRangef, GL_DEPTH_RANGE);
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (DepthClamp, GL_DEPTH_CLAMP);
+	ATTRIBUTE_TYPE_DEFS            (DepthMask, GL_DEPTH_WRITEMASK);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (StencilTest, GL_STENCIL_TEST);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (AlphaTest, GL_ALPHA_TEST);
+	ATTRIBUTE_TYPE_DEFS            (AlphaFunc, GL_ALPHA_TEST_FUNC, GL_ALPHA_TEST_REF);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (Blending, GL_BLEND);
+	ATTRIBUTE_TYPE_DEFS            (BlendFunc, GL_BLEND_SRC, GL_BLEND_DST);
+	ATTRIBUTE_TYPE_DEFS            (BlendColor, GL_BLEND_COLOR);
+	ATTRIBUTE_TYPE_DEFS            (ColorMask, GL_COLOR_WRITEMASK);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (PrimitiveRestart, GL_PRIMITIVE_RESTART);
+	ATTRIBUTE_TYPE_DEFS            (PrimitiveRestartIndex, GL_PRIMITIVE_RESTART_INDEX);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (Multisampling, GL_MULTISAMPLE);
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (AlphaToCoverage, GL_SAMPLE_ALPHA_TO_COVERAGE);
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (AlphaToOne, GL_SAMPLE_ALPHA_TO_ONE);
+
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (CubemapSeamless, GL_TEXTURE_CUBE_MAP_SEAMLESS);
+	CAPABILITY_ATTRIBUTE_TYPE_DEFS (PointSize, GL_PROGRAM_POINT_SIZE);
+
+	extern std::tuple<
+		ATTRIBUTE(PolygonMode),
+		ATTRIBUTE(PolygonOffsetFill),
+		ATTRIBUTE(PolygonOffsetLine),
+		ATTRIBUTE(PolygonOffsetPoint),
+		ATTRIBUTE(LineWidth),
+		ATTRIBUTE(Viewport),
+		ATTRIBUTE(FrontFace),
+		ATTRIBUTE(Culling),
+		ATTRIBUTE(CullFace),
+		ATTRIBUTE(ClipDistance)<0>,
+		ATTRIBUTE(ClipDistance)<1>,
+		ATTRIBUTE(ClipDistance)<2>,
+		ATTRIBUTE(ClipDistance)<3>,
+		ATTRIBUTE(ClipDistance)<4>,
+		ATTRIBUTE(ClipDistance)<5>,
+		ATTRIBUTE(ClipDistance)<6>,
+		ATTRIBUTE(ClipDistance)<7>,
+		ATTRIBUTE(ScissorTest),
+		ATTRIBUTE(Scissor),
+		ATTRIBUTE(DepthTest),
+		ATTRIBUTE(DepthFunc),
+		ATTRIBUTE(DepthRangef),
+		ATTRIBUTE(DepthClamp),
+		ATTRIBUTE(DepthMask),
+		ATTRIBUTE(StencilTest),
+		ATTRIBUTE(AlphaTest),
+		ATTRIBUTE(AlphaFunc),
+		ATTRIBUTE(Blending),
+		ATTRIBUTE(BlendFunc),
+		ATTRIBUTE(BlendColor),
+		ATTRIBUTE(ColorMask),
+		ATTRIBUTE(PrimitiveRestart),
+		ATTRIBUTE(PrimitiveRestartIndex),
+		ATTRIBUTE(Multisampling),
+		ATTRIBUTE(AlphaToCoverage),
+		ATTRIBUTE(AlphaToOne),
+		ATTRIBUTE(CubemapSeamless),
+		ATTRIBUTE(PointSize)
+	> Attributes;
```

Comment:
I'm not entirely sure why this one is declared as `extern`?

Reply:
Ah I take it back, it's a global variable.

Any objections against moving `namespace State` --> `struct State` and make `extern Attributes` `inline static Attributes;`?

Reply:
If you do not declare a variable extern, it will be duplicated for each unit that uses it. _Extern_ qualifier ensures that only 1 instance of Attributes will be created.

Reply:
> Ah I take it back, it's a global variable.
> 
> Any objections against moving `namespace State` --> `struct State` and make `extern Attributes` `inline static Attributes;`?

Yes, to both.
First of all, making State a struct serves no purpose. It makes impossible to use namespace and shorter names for each Attribute (remember what I mentioned, GL::State::DepthMask is a long writing).
Explained why static would be incorrect above^.

Reply:
Many other occasions we are having global static variables across Spring code, I assume, are incorrect as well.
Every global variable, if it is meant to be instantiated only once, should have _extern_ qualifier.